JaCoCo Instrumentation Failure with material-kolor 4.0.3

## Summary
Upgrading from material-kolor 3.0.1 to 4.0.3 causes build failures when JaCoCo code coverage is enabled for debug builds. The error occurs during the JaCoCo transformation phase with a `MethodTooLargeException`.


   ## Environment

   - **material-kolor version**: 4.0.3
   - **Gradle version**: 9.1.0
   - **AGP (Android Gradle Plugin)**: 8.13.0
   - **JaCoCo version**: 0.8.13 (also tested with 0.8.14, same issue)
   - **Kotlin**: 2.2.0
   - **JDK**: 21.0.3 (Azul Systems)
   - **OS**: macOS 15.6.1 (aarch64)
   - **Min SDK**: 23
   - **Target SDK**: 36

   ## Error Details

   ### Build Task Failure

   ```
   Execution failed for task ':app-compose:mergeExtDexFossDebug'.
   > Could not resolve all files for configuration ':app-compose:fossDebugRuntimeClasspath'.
      > Failed to transform material-kolor-debug.aar (com.materialkolor:material-kolor-android-debug:4.0.3) to match attributes {artifactType=android-dex,
   com.android.build.api.attributes.BuildTypeAttr=debug, dexing-component-attributes=ComponentSpecificParameters(minSdkVersion=23, debuggable=true,enableCoreLibraryDesugaring=true, enableGlobalSynthetics=true, enableApiModeling=false, dependenciesClassesAreInstrumented=false, asmTransformComponent=null,useJacocoTransformInstrumentation=true, enableDesugaring=true, needsClasspath=true, useFullClasspath=false, componentIfUsingFullClasspath=null),org.gradle.category=library, org.gradle.jvm.environment=android, org.gradle.status=release, org.gradle.usage=java-runtime, org.jetbrains.kotlin.platform.type=androidJvm}.
         > Execution failed for JacocoTransform:
   .gradle/caches/9.1.0/transforms/471c60d28eb48fcdd5010542f09f36fd/transformed/material-kolor-debug-runtime.jar.
   ```

### Root Cause

   The complete stack trace reveals the actual issue:

   ```
   Caused by: java.io.IOException: Error while instrumenting com/materialkolor/ktx/ColorSchemeKt.class with JaCoCo 0.8.13.202504020838/78d5eff.
   	at org.jacoco.core.instr.Instrumenter.instrumentError(Instrumenter.java:161)
   	at org.jacoco.core.instr.Instrumenter.instrument(Instrumenter.java:111)
   	at org.jacoco.core.instr.Instrumenter.instrument(Instrumenter.java:136)
   	... 119 more
   Caused by: org.objectweb.asm.MethodTooLargeException: Method too large: com/materialkolor/ktx/ColorSchemeKt.animateColorScheme (Landroidx/compose/material3/ColorScheme;Lkotlin/jvm/functions/Function3;Ljava/lang/String;Landroidx/compose/runtime/Composer;II)Landroidx/compose/material3/ColorScheme;
   	at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2088)
   	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:506)
   	at org.jacoco.core.instr.Instrumenter.instrument(Instrumenter.java:92)
   	at org.jacoco.core.instr.Instrumenter.instrument(Instrumenter.java:109)
   	... 120 more
   ```

   **The problematic method**: `com/materialkolor/ktx/ColorSchemeKt.animateColorScheme`

https://github.com/lihenggui/blocker/actions/runs/19244555472/job/55015234801

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -50,3 +50,6 @@ jobs:
 
       - name: Upstream Tests
         run: ./gradlew :mcu-upstream:test --scan
+
+      - name: Material Kolor Tests
+        run: ./gradlew :material-kolor:jvmTest --scan

--- a/material-kolor/build.gradle.kts
+++ b/material-kolor/build.gradle.kts
@@ -62,6 +62,12 @@ kotlin {
 
         commonTest.dependencies {
             implementation(kotlin("test"))
+            @OptIn(org.jetbrains.compose.ExperimentalComposeLibrary::class)
+            implementation(compose.uiTest)
+        }
+
+        jvmTest.dependencies {
+            implementation(compose.desktop.currentOs)
         }
     }
 

--- a/material-kolor/src/commonMain/kotlin/com/materialkolor/ktx/ColorScheme.kt
+++ b/material-kolor/src/commonMain/kotlin/com/materialkolor/ktx/ColorScheme.kt
@@ -29,294 +29,377 @@ public fun animateColorScheme(
 ): ColorScheme {
     val transition = updateTransition(targetState = colorScheme, label = label)
 
-    val primary by transition.animateColor(
+    val primary = transition.animatePrimaryColors(animationSpec)
+    val secondary = transition.animateSecondaryColors(animationSpec)
+    val tertiary = transition.animateTertiaryColors(animationSpec)
+    val surface = transition.animateSurfaceColors(animationSpec)
+    val backgroundError = transition.animateBackgroundAndErrorColors(animationSpec)
+    val utility = transition.animateUtilityColors(animationSpec)
+    val fixed = transition.animateFixedColors(animationSpec)
+
+    return colorScheme.copy(
+        primary = primary[0],
+        onPrimary = primary[1],
+        primaryContainer = primary[2],
+        onPrimaryContainer = primary[3],
+        inversePrimary = primary[4],
+        secondary = secondary[0],
+        onSecondary = secondary[1],
+        secondaryContainer = secondary[2],
+        onSecondaryContainer = secondary[3],
+        tertiary = tertiary[0],
+        onTertiary = tertiary[1],
+        tertiaryContainer = tertiary[2],
+        onTertiaryContainer = tertiary[3],
+        background = backgroundError[0],
+        onBackground = backgroundError[1],
+        surface = surface[0],
+        onSurface = surface[1],
+        surfaceVariant = surface[2],
+        onSurfaceVariant = surface[3],
+        surfaceTint = surface[4],
+        inverseSurface = utility[0],
+        inverseOnSurface = utility[1],
+        error = backgroundError[2],
+        onError = backgroundError[3],
+        errorContainer = backgroundError[4],
+        onErrorContainer = backgroundError[5],
+        outline = utility[2],
+        outlineVariant = utility[3],
+        scrim = utility[4],
+        surfaceBright = surface[5],
+        surfaceDim = surface[6],
+        surfaceContainer = surface[7],
+        surfaceContainerHigh = surface[8],
+        surfaceContainerHighest = surface[9],
+        surfaceContainerLow = surface[10],
+        surfaceContainerLowest = surface[11],
+        primaryFixed = fixed[0],
+        primaryFixedDim = fixed[1],
+        onPrimaryFixed = fixed[2],
+        onPrimaryFixedVariant = fixed[3],
+        secondaryFixed = fixed[4],
+        secondaryFixedDim = fixed[5],
+        onSecondaryFixed = fixed[6],
+        onSecondaryFixedVariant = fixed[7],
+        tertiaryFixed = fixed[8],
+        tertiaryFixedDim = fixed[9],
+        onTertiaryFixed = fixed[10],
+        onTertiaryFixedVariant = fixed[11],
+    )
+}
+
+@Composable
+private fun Transition<ColorScheme>.animatePrimaryColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val primary by animateColor(
         label = "color_primary",
         targetValueByState = { it.primary },
         transitionSpec = animationSpec,
     )
-    val primaryContainer by transition.animateColor(
-        label = "color_primaryContainer",
-        targetValueByState = { it.primaryContainer },
-        transitionSpec = animationSpec,
-    )
-    val secondary by transition.animateColor(
-        label = "color_secondary",
-        targetValueByState = { it.secondary },
-        transitionSpec = animationSpec,
-    )
-    val secondaryContainer by transition.animateColor(
-        label = "color_secondaryContainer",
-        targetValueByState = { it.secondaryContainer },
-        transitionSpec = animationSpec,
-    )
-    val tertiary by transition.animateColor(
-        label = "color_tertiary",
-        targetValueByState = { it.tertiary },
-        transitionSpec = animationSpec,
-    )
-    val tertiaryContainer by transition.animateColor(
-        label = "color_tertiaryContainer",
-        targetValueByState = { it.tertiaryContainer },
-        transitionSpec = animationSpec,
-    )
-    val background by transition.animateColor(
-        label = "color_background",
-        targetValueByState = { it.background },
-        transitionSpec = animationSpec,
-    )
-    val surface by transition.animateColor(
-        label = "color_surface",
-        targetValueByState = { it.surface },
-        transitionSpec = animationSpec,
-    )
-    val surfaceTint by transition.animateColor(
-        label = "color_surfaceTint",
-        targetValueByState = { it.surfaceTint },
-        transitionSpec = animationSpec,
-    )
-    val surfaceBright by transition.animateColor(
-        label = "color_surfaceBright",
-        targetValueByState = { it.surfaceBright },
-        transitionSpec = animationSpec,
-    )
-    val surfaceDim by transition.animateColor(
-        label = "color_surfaceDim",
-        targetValueByState = { it.surfaceDim },
-        transitionSpec = animationSpec,
-    )
-    val surfaceContainer by transition.animateColor(
-        label = "color_surfaceContainer",
-        targetValueByState = { it.surfaceContainer },
-        transitionSpec = animationSpec,
-    )
-    val surfaceContainerHigh by transition.animateColor(
-        label = "color_surfaceContainerHigh",
-        targetValueByState = { it.surfaceContainerHigh },
-        transitionSpec = animationSpec,
-    )
-    val surfaceContainerHighest by transition.animateColor(
-        label = "color_surfaceContainerHighest",
-        targetValueByState = { it.surfaceContainerHighest },
-        transitionSpec = animationSpec,
-    )
-    val surfaceContainerLow by transition.animateColor(
-        label = "color_surfaceContainerLow",
-        targetValueByState = { it.surfaceContainerLow },
-        transitionSpec = animationSpec,
-    )
-    val surfaceContainerLowest by transition.animateColor(
-        label = "color_surfaceContainerLowest",
-        targetValueByState = { it.surfaceContainerLowest },
-        transitionSpec = animationSpec,
-    )
-    val surfaceVariant by transition.animateColor(
-        label = "color_surfaceVariant",
-        targetValueByState = { it.surfaceVariant },
-        transitionSpec = animationSpec,
-    )
-    val error by transition.animateColor(
-        label = "color_error",
-        targetValueByState = { it.error },
-        transitionSpec = animationSpec,
-    )
-    val errorContainer by transition.animateColor(
-        label = "color_errorContainer",
-        targetValueByState = { it.errorContainer },
-        transitionSpec = animationSpec,
-    )
-    val onPrimary by transition.animateColor(
+    val onPrimary by animateColor(
         label = "color_onPrimary",
         targetValueByState = { it.onPrimary },
         transitionSpec = animationSpec,
     )
-    val onPrimaryContainer by transition.animateColor(
+    val primaryContainer by animateColor(
+        label = "color_primaryContainer",
+        targetValueByState = { it.primaryContainer },
+        transitionSpec = animationSpec,
+    )
+    val onPrimaryContainer by animateColor(
         label = "color_onPrimaryContainer",
         targetValueByState = { it.onPrimaryContainer },
         transitionSpec = animationSpec,
     )
-    val onSecondary by transition.animateColor(
-        label = "color_onSecondary",
-        targetValueByState = { it.onSecondary },
-        transitionSpec = animationSpec,
-    )
-    val onSecondaryContainer by transition.animateColor(
-        label = "color_onSecondaryContainer",
-        targetValueByState = { it.onSecondaryContainer },
-        transitionSpec = animationSpec,
-    )
-    val onTertiary by transition.animateColor(
-        label = "color_onTertiary",
-        targetValueByState = { it.onTertiary },
-        transitionSpec = animationSpec,
-    )
-    val onTertiaryContainer by transition.animateColor(
-        label = "color_onTertiaryContainer",
-        targetValueByState = { it.onTertiaryContainer },
-        transitionSpec = animationSpec,
-    )
-    val onBackground by transition.animateColor(
-        label = "color_onBackground",
-        targetValueByState = { it.onBackground },
-        transitionSpec = animationSpec,
-    )
-    val onSurface by transition.animateColor(
-        label = "color_onSurface",
-        targetValueByState = { it.onSurface },
-        transitionSpec = animationSpec,
-    )
-    val onSurfaceVariant by transition.animateColor(
-        label = "color_onSurfaceVariant",
-        targetValueByState = { it.onSurfaceVariant },
-        transitionSpec = animationSpec,
-    )
-    val onError by transition.animateColor(
-        label = "color_onError",
-        targetValueByState = { it.onError },
-        transitionSpec = animationSpec,
-    )
-    val onErrorContainer by transition.animateColor(
-        label = "color_onErrorContainer",
-        targetValueByState = { it.onErrorContainer },
-        transitionSpec = animationSpec,
-    )
-    val inversePrimary by transition.animateColor(
+    val inversePrimary by animateColor(
         label = "color_inversePrimary",
         targetValueByState = { it.inversePrimary },
         transitionSpec = animationSpec,
     )
-    val inverseSurface by transition.animateColor(
+    return listOf(primary, onPrimary, primaryContainer, onPrimaryContainer, inversePrimary)
+}
+
+@Composable
+private fun Transition<ColorScheme>.animateSecondaryColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val secondary by animateColor(
+        label = "color_secondary",
+        targetValueByState = { it.secondary },
+        transitionSpec = animationSpec,
+    )
+    val onSecondary by animateColor(
+        label = "color_onSecondary",
+        targetValueByState = { it.onSecondary },
+        transitionSpec = animationSpec,
+    )
+    val secondaryContainer by animateColor(
+        label = "color_secondaryContainer",
+        targetValueByState = { it.secondaryContainer },
+        transitionSpec = animationSpec,
+    )
+    val onSecondaryContainer by animateColor(
+        label = "color_onSecondaryContainer",
+        targetValueByState = { it.onSecondaryContainer },
+        transitionSpec = animationSpec,
+    )
+    return listOf(secondary, onSecondary, secondaryContainer, onSecondaryContainer)
+}
+
+@Composable
+private fun Transition<ColorScheme>.animateTertiaryColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val tertiary by animateColor(
+        label = "color_tertiary",
+        targetValueByState = { it.tertiary },
+        transitionSpec = animationSpec,
+    )
+    val onTertiary by animateColor(
+        label = "color_onTertiary",
+        targetValueByState = { it.onTertiary },
+        transitionSpec = animationSpec,
+    )
+    val tertiaryContainer by animateColor(
+        label = "color_tertiaryContainer",
+        targetValueByState = { it.tertiaryContainer },
+        transitionSpec = animationSpec,
+    )
+    val onTertiaryContainer by animateColor(
+        label = "color_onTertiaryContainer",
+        targetValueByState = { it.onTertiaryContainer },
+        transitionSpec = animationSpec,
+    )
+    return listOf(tertiary, onTertiary, tertiaryContainer, onTertiaryContainer)
+}
+
+@Composable
+private fun Transition<ColorScheme>.animateSurfaceColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val surface by animateColor(
+        label = "color_surface",
+        targetValueByState = { it.surface },
+        transitionSpec = animationSpec,
+    )
+    val onSurface by animateColor(
+        label = "color_onSurface",
+        targetValueByState = { it.onSurface },
+        transitionSpec = animationSpec,
+    )
+    val surfaceVariant by animateColor(
+        label = "color_surfaceVariant",
+        targetValueByState = { it.surfaceVariant },
+        transitionSpec = animationSpec,
+    )
+    val onSurfaceVariant by animateColor(
+        label = "color_onSurfaceVariant",
+        targetValueByState = { it.onSurfaceVariant },
+        transitionSpec = animationSpec,
+    )
+    val surfaceTint by animateColor(
+        label = "color_surfaceTint",
+        targetValueByState = { it.surfaceTint },
+        transitionSpec = animationSpec,
+    )
+    val surfaceBright by animateColor(
+        label = "color_surfaceBright",
+        targetValueByState = { it.surfaceBright },
+        transitionSpec = animationSpec,
+    )
+    val surfaceDim by animateColor(
+        label = "color_surfaceDim",
+        targetValueByState = { it.surfaceDim },
+        transitionSpec = animationSpec,
+    )
+    val surfaceContainer by animateColor(
+        label = "color_surfaceContainer",
+        targetValueByState = { it.surfaceContainer },
+        transitionSpec = animationSpec,
+    )
+    val surfaceContainerHigh by animateColor(
+        label = "color_surfaceContainerHigh",
+        targetValueByState = { it.surfaceContainerHigh },
+        transitionSpec = animationSpec,
+    )
+    val surfaceContainerHighest by animateColor(
+        label = "color_surfaceContainerHighest",
+        targetValueByState = { it.surfaceContainerHighest },
+        transitionSpec = animationSpec,
+    )
+    val surfaceContainerLow by animateColor(
+        label = "color_surfaceContainerLow",
+        targetValueByState = { it.surfaceContainerLow },
+        transitionSpec = animationSpec,
+    )
+    val surfaceContainerLowest by animateColor(
+        label = "color_surfaceContainerLowest",
+        targetValueByState = { it.surfaceContainerLowest },
+        transitionSpec = animationSpec,
+    )
+    return listOf(
+        surface,
+        onSurface,
+        surfaceVariant,
+        onSurfaceVariant,
+        surfaceTint,
+        surfaceBright,
+        surfaceDim,
+        surfaceContainer,
+        surfaceContainerHigh,
+        surfaceContainerHighest,
+        surfaceContainerLow,
+        surfaceContainerLowest,
+    )
+}
+
+@Composable
+private fun Transition<ColorScheme>.animateBackgroundAndErrorColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val background by animateColor(
+        label = "color_background",
+        targetValueByState = { it.background },
+        transitionSpec = animationSpec,
+    )
+    val onBackground by animateColor(
+        label = "color_onBackground",
+        targetValueByState = { it.onBackground },
+        transitionSpec = animationSpec,
+    )
+    val error by animateColor(
+        label = "color_error",
+        targetValueByState = { it.error },
+        transitionSpec = animationSpec,
+    )
+    val onError by animateColor(
+        label = "color_onError",
+        targetValueByState = { it.onError },
+        transitionSpec = animationSpec,
+    )
+    val errorContainer by animateColor(
+        label = "color_errorContainer",
+        targetValueByState = { it.errorContainer },
+        transitionSpec = animationSpec,
+    )
+    val onErrorContainer by animateColor(
+        label = "color_onErrorContainer",
+        targetValueByState = { it.onErrorContainer },
+        transitionSpec = animationSpec,
+    )
+    return listOf(background, onBackground, error, onError, errorContainer, onErrorContainer)
+}
+
+@Composable
+private fun Transition<ColorScheme>.animateUtilityColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val inverseSurface by animateColor(
         label = "color_inverseSurface",
         targetValueByState = { it.inverseSurface },
         transitionSpec = animationSpec,
     )
-    val inverseOnSurface by transition.animateColor(
+    val inverseOnSurface by animateColor(
         label = "color_inverseOnSurface",
         targetValueByState = { it.inverseOnSurface },
         transitionSpec = animationSpec,
     )
-    val outline by transition.animateColor(
+    val outline by animateColor(
         label = "color_outline",
         targetValueByState = { it.outline },
         transitionSpec = animationSpec,
     )
-    val outlineVariant by transition.animateColor(
+    val outlineVariant by animateColor(
         label = "color_outlineVariant",
         targetValueByState = { it.outlineVariant },
         transitionSpec = animationSpec,
     )
-    val scrim by transition.animateColor(
+    val scrim by animateColor(
         label = "color_scrim",
         targetValueByState = { it.scrim },
         transitionSpec = animationSpec,
     )
-    val primaryFixed by transition.animateColor(
+    return listOf(inverseSurface, inverseOnSurface, outline, outlineVariant, scrim)
+}
+
+@Composable
+private fun Transition<ColorScheme>.animateFixedColors(
+    animationSpec: @Composable Transition.Segment<ColorScheme>.() -> FiniteAnimationSpec<Color>,
+): List<Color> {
+    val primaryFixed by animateColor(
         label = "color_primaryFixed",
         targetValueByState = { it.primaryFixed },
         transitionSpec = animationSpec,
     )
-    val primaryFixedDim by transition.animateColor(
+    val primaryFixedDim by animateColor(
         label = "color_primaryFixedDim",
         targetValueByState = { it.primaryFixedDim },
         transitionSpec = animationSpec,
     )
-    val onPrimaryFixed by transition.animateColor(
+    val onPrimaryFixed by animateColor(
         label = "color_onPrimaryFixed",
         targetValueByState = { it.onPrimaryFixed },
         transitionSpec = animationSpec,
     )
-    val onPrimaryFixedVariant by transition.animateColor(
+    val onPrimaryFixedVariant by animateColor(
         label = "color_onPrimaryFixedVariant",
         targetValueByState = { it.onPrimaryFixedVariant },
         transitionSpec = animationSpec,
     )
-    val secondaryFixed by transition.animateColor(
+    val secondaryFixed by animateColor(
         label = "color_secondaryFixed",
         targetValueByState = { it.secondaryFixed },
         transitionSpec = animationSpec,
     )
-    val secondaryFixedDim by transition.animateColor(
+    val secondaryFixedDim by animateColor(
         label = "color_secondaryFixedDim",
         targetValueByState = { it.secondaryFixedDim },
         transitionSpec = animationSpec,
     )
-    val onSecondaryFixed by transition.animateColor(
+    val onSecondaryFixed by animateColor(
         label = "color_onSecondaryFixed",
         targetValueByState = { it.onSecondaryFixed },
         transitionSpec = animationSpec,
     )
-    val onSecondaryFixedVariant by transition.animateColor(
+    val onSecondaryFixedVariant by animateColor(
         label = "color_onSecondaryFixedVariant",
         targetValueByState = { it.onSecondaryFixedVariant },
         transitionSpec = animationSpec,
     )
-    val tertiaryFixed by transition.animateColor(
+    val tertiaryFixed by animateColor(
         label = "color_tertiaryFixed",
         targetValueByState = { it.tertiaryFixed },
         transitionSpec = animationSpec,
     )
-    val tertiaryFixedDim by transition.animateColor(
+    val tertiaryFixedDim by animateColor(
         label = "color_tertiaryFixedDim",
         targetValueByState = { it.tertiaryFixedDim },
         transitionSpec = animationSpec,
     )
-    val onTertiaryFixed by transition.animateColor(
+    val onTertiaryFixed by animateColor(
         label = "color_onTertiaryFixed",
         targetValueByState = { it.onTertiaryFixed },
         transitionSpec = animationSpec,
     )
-    val onTertiaryFixedVariant by transition.animateColor(
+    val onTertiaryFixedVariant by animateColor(
         label = "color_onTertiaryFixedVariant",
         targetValueByState = { it.onTertiaryFixedVariant },
         transitionSpec = animationSpec,
     )
-    return colorScheme.copy(
-        primary = primary,
-        onPrimary = onPrimary,
-        primaryContainer = primaryContainer,
-        onPrimaryContainer = onPrimaryContainer,
-        inversePrimary = inversePrimary,
-        secondary = secondary,
-        onSecondary = onSecondary,
-        secondaryContainer = secondaryContainer,
-        onSecondaryContainer = onSecondaryContainer,
-        tertiary = tertiary,
-        onTertiary = onTertiary,
-        tertiaryContainer = tertiaryContainer,
-        onTertiaryContainer = onTertiaryContainer,
-        background = background,
-        onBackground = onBackground,
-        surface = surface,
-        onSurface = onSurface,
-        surfaceVariant = surfaceVariant,
-        onSurfaceVariant = onSurfaceVariant,
-        surfaceTint = surfaceTint,
-        inverseSurface = inverseSurface,
-        inverseOnSurface = inverseOnSurface,
-        error = error,
-        onError = onError,
-        errorContainer = errorContainer,
-        onErrorContainer = onErrorContainer,
-        outline = outline,
-        outlineVariant = outlineVariant,
-        scrim = scrim,
-        surfaceBright = surfaceBright,
-        surfaceDim = surfaceDim,
-        surfaceContainer = surfaceContainer,
-        surfaceContainerHigh = surfaceContainerHigh,
-        surfaceContainerHighest = surfaceContainerHighest,
-        surfaceContainerLow = surfaceContainerLow,
-        surfaceContainerLowest = surfaceContainerLowest,
-        primaryFixed = primaryFixed,
-        primaryFixedDim = primaryFixedDim,
-        onPrimaryFixed = onPrimaryFixed,
-        onPrimaryFixedVariant = onPrimaryFixedVariant,
-        secondaryFixed = secondaryFixed,
-        secondaryFixedDim = secondaryFixedDim,
-        onSecondaryFixed = onSecondaryFixed,
-        onSecondaryFixedVariant = onSecondaryFixedVariant,
-        tertiaryFixed = tertiaryFixed,
-        tertiaryFixedDim = tertiaryFixedDim,
-        onTertiaryFixed = onTertiaryFixed,
-        onTertiaryFixedVariant = onTertiaryFixedVariant,
+    return listOf(
+        primaryFixed,
+        primaryFixedDim,
+        onPrimaryFixed,
+        onPrimaryFixedVariant,
+        secondaryFixed,
+        secondaryFixedDim,
+        onSecondaryFixed,
+        onSecondaryFixedVariant,
+        tertiaryFixed,
+        tertiaryFixedDim,
+        onTertiaryFixed,
+        onTertiaryFixedVariant,
     )
 }

--- a/material-kolor/src/jvmTest/kotlin/com/materialkolor/ktx/ColorSchemeTest.kt
+++ b/material-kolor/src/jvmTest/kotlin/com/materialkolor/ktx/ColorSchemeTest.kt
@@ -1,0 +1,226 @@
+package com.materialkolor.ktx
+
+import androidx.compose.material3.ColorScheme
+import androidx.compose.material3.darkColorScheme
+import androidx.compose.material3.lightColorScheme
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.test.ExperimentalTestApi
+import androidx.compose.ui.test.runComposeUiTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+
+@OptIn(ExperimentalTestApi::class)
+class ColorSchemeTest {
+    private val testLightScheme = lightColorScheme(
+        primary = Color.Red,
+        secondary = Color.Blue,
+        tertiary = Color.Green,
+        surface = Color.White,
+        background = Color(0xFFFFFBFE),
+    )
+
+    private val testDarkScheme = darkColorScheme(
+        primary = Color.Cyan,
+        secondary = Color.Magenta,
+        tertiary = Color.Yellow,
+        surface = Color.Black,
+        background = Color(0xFF1C1B1F),
+    )
+
+    @Test
+    fun animateColorScheme_preservesAllPrimaryColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.primary, result.primary)
+        assertEquals(testLightScheme.onPrimary, result.onPrimary)
+        assertEquals(testLightScheme.primaryContainer, result.primaryContainer)
+        assertEquals(testLightScheme.onPrimaryContainer, result.onPrimaryContainer)
+        assertEquals(testLightScheme.inversePrimary, result.inversePrimary)
+    }
+
+    @Test
+    fun animateColorScheme_preservesSecondaryColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.secondary, result.secondary)
+        assertEquals(testLightScheme.onSecondary, result.onSecondary)
+        assertEquals(testLightScheme.secondaryContainer, result.secondaryContainer)
+        assertEquals(testLightScheme.onSecondaryContainer, result.onSecondaryContainer)
+    }
+
+    @Test
+    fun animateColorScheme_preservesTertiaryColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.tertiary, result.tertiary)
+        assertEquals(testLightScheme.onTertiary, result.onTertiary)
+        assertEquals(testLightScheme.tertiaryContainer, result.tertiaryContainer)
+        assertEquals(testLightScheme.onTertiaryContainer, result.onTertiaryContainer)
+    }
+
+    @Test
+    fun animateColorScheme_preservesSurfaceColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.surface, result.surface)
+        assertEquals(testLightScheme.onSurface, result.onSurface)
+        assertEquals(testLightScheme.surfaceVariant, result.surfaceVariant)
+        assertEquals(testLightScheme.onSurfaceVariant, result.onSurfaceVariant)
+        assertEquals(testLightScheme.surfaceTint, result.surfaceTint)
+        assertEquals(testLightScheme.surfaceBright, result.surfaceBright)
+        assertEquals(testLightScheme.surfaceDim, result.surfaceDim)
+        assertEquals(testLightScheme.surfaceContainer, result.surfaceContainer)
+        assertEquals(testLightScheme.surfaceContainerHigh, result.surfaceContainerHigh)
+        assertEquals(testLightScheme.surfaceContainerHighest, result.surfaceContainerHighest)
+        assertEquals(testLightScheme.surfaceContainerLow, result.surfaceContainerLow)
+        assertEquals(testLightScheme.surfaceContainerLowest, result.surfaceContainerLowest)
+    }
+
+    @Test
+    fun animateColorScheme_preservesErrorAndBackgroundColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.background, result.background)
+        assertEquals(testLightScheme.onBackground, result.onBackground)
+        assertEquals(testLightScheme.error, result.error)
+        assertEquals(testLightScheme.onError, result.onError)
+        assertEquals(testLightScheme.errorContainer, result.errorContainer)
+        assertEquals(testLightScheme.onErrorContainer, result.onErrorContainer)
+    }
+
+    @Test
+    fun animateColorScheme_preservesUtilityColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.inverseSurface, result.inverseSurface)
+        assertEquals(testLightScheme.inverseOnSurface, result.inverseOnSurface)
+        assertEquals(testLightScheme.outline, result.outline)
+        assertEquals(testLightScheme.outlineVariant, result.outlineVariant)
+        assertEquals(testLightScheme.scrim, result.scrim)
+    }
+
+    @Test
+    fun animateColorScheme_preservesFixedColors() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testLightScheme,
+                label = "test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.primaryFixed, result.primaryFixed)
+        assertEquals(testLightScheme.primaryFixedDim, result.primaryFixedDim)
+        assertEquals(testLightScheme.onPrimaryFixed, result.onPrimaryFixed)
+        assertEquals(testLightScheme.onPrimaryFixedVariant, result.onPrimaryFixedVariant)
+        assertEquals(testLightScheme.secondaryFixed, result.secondaryFixed)
+        assertEquals(testLightScheme.secondaryFixedDim, result.secondaryFixedDim)
+        assertEquals(testLightScheme.onSecondaryFixed, result.onSecondaryFixed)
+        assertEquals(testLightScheme.onSecondaryFixedVariant, result.onSecondaryFixedVariant)
+        assertEquals(testLightScheme.tertiaryFixed, result.tertiaryFixed)
+        assertEquals(testLightScheme.tertiaryFixedDim, result.tertiaryFixedDim)
+        assertEquals(testLightScheme.onTertiaryFixed, result.onTertiaryFixed)
+        assertEquals(testLightScheme.onTertiaryFixedVariant, result.onTertiaryFixedVariant)
+    }
+
+    @Test
+    fun animateColorScheme_handlesColorSchemeTransition() = runComposeUiTest {
+        var colorScheme by mutableStateOf(testLightScheme)
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = colorScheme,
+                label = "transition_test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testLightScheme.primary, result.primary)
+
+        colorScheme = testDarkScheme
+        waitForIdle()
+
+        assertNotNull(result)
+    }
+
+    @Test
+    fun animateColorScheme_worksWithDarkColorScheme() = runComposeUiTest {
+        var result: ColorScheme? = null
+
+        setContent {
+            result = animateColorScheme(
+                colorScheme = testDarkScheme,
+                label = "dark_test",
+            )
+        }
+
+        waitForIdle()
+        assertNotNull(result)
+        assertEquals(testDarkScheme.primary, result.primary)
+        assertEquals(testDarkScheme.secondary, result.secondary)
+        assertEquals(testDarkScheme.tertiary, result.tertiary)
+    }
+}

